File: test/strategy/bench_aip.py
#!/usr/bin/env python3

import time
import datetime
import eatmud


def bench_aip(trans: eatmud.Transaction) -> list[list[float]]:
    results = []
    for save_log in (True, False,):
        for save_record in (True, False):
            name = f'save_log={save_log}, save_record={save_record}'
            now = time.time()
            res = []
            for day in range(1, 29):
                it = trans.iter(save_log, save_record)
                eatmud.strategy.aip_monthly(it, day, [1000.]*2, [0.]*2)
                res.append(it.asset())
            total_time = time.time() - now
            print(f'running aip({name}) took '
                  f'{total_time*1000} milli seconds')
            results.append(res)
    return results


def test_aip():
    hs300 = eatmud.Fund.from_stock(eatmud.io.read_tdx("tdx/test-hs300.txt"))
    gz2000 = eatmud.Fund.from_stock(eatmud.io.read_tdx("tdx/test-gz2000.txt"))
    start_date = datetime.datetime.strptime('20110101', '%Y%m%d').date()
    end_date = datetime.datetime.strptime('20240101', '%Y%m%d').date()
    trans = eatmud.Transaction([hs300, gz2000], start_date, end_date)
    results = bench_aip(trans)
    print(results)


if __name__ == '__main__':
    test_aip()

File: eatmud_core/src/bin/bench_aip.rs
//use eatmud;

use std::time::Instant;

use chrono::NaiveDate;
use eatmud::{io::read_tdx, Fund, Transaction};

fn bench_aip(trans: &Transaction) -> Vec<Vec<f64>> {
    let mut results = Vec::new();
    for save_log in [true, false] {
        for save_record in [true, false] {
            let name = format!("save_log={}, save_record={}", save_log, save_record);
            let now = Instant::now();
            let mut res = Vec::new();
            for day in 1..29 {
                let mut it = trans.iter(save_log, save_record);
                eatmud::strategy::aip_monthly(&mut it, day, &[1000., 1000.], &[0., 0.]).unwrap();
                res.push(it.asset());
            }
            let total_time = Instant::now() - now;
            println!(
                "running aip({}) took {} milli seconds",
                name,
                total_time.as_micros() as f64 / 1000.
            );
            results.push(res);
        }
    }
    results
}

fn main() {
    let hs300 = Fund::from(&read_tdx("tdx/test-hs300.txt").unwrap());
    let gz2000 = Fund::from(&read_tdx("tdx/test-gz2000.txt").unwrap());
    let start_date = NaiveDate::parse_from_str("20110101", "%Y%m%d").unwrap();
    let end_date = NaiveDate::parse_from_str("20240101", "%Y%m%d").unwrap();
    let trans = Transaction::new(&[&hs300, &gz2000], Some(start_date), Some(end_date));
    let results = bench_aip(&trans);
    println!("{:?}", results);
    assert_eq!(results[0], results[1]);
    assert_eq!(results[0], results[2]);
    assert_eq!(results[0], results[3]);
}
